### ISSUE TITLE
Documentation example request

### ISSUE THREAD

**Issue by theboomer** [SpongePowered/SpongeDocs#371]
If requests may infact be made...

gabizou and the others are very good at helping with dumb questions, or questions that need some additional information to move forward in the right direction, but, as their time is also very limited it becomes cumbersome and undesireable to keep going back for 'please clarify that part' explanations, yet, often when a good, solid working n-line code example is given and able to be inspected, the ah-hah breaks through, and opens the door on clarity and understanding finally, allowing exporation, changes, and basically having a prototype in hands to dismantle and put together.
Something permeating the forum pages for ages has been custom data storage on a tile or entity, but its always rather unclear (in past due to due repeated changes to things altering whatever was written as 'you will need to do.." primarily).  The notion is still unclear when its summarized with a handwave 'you need to blah with your custom data class and blah custom thing then blah an immutable blah then blah mutable version blah and that is all you need."

I say a handwave because its not obvious to anyone not tuned-in to the thought processes of it, what those 'you supply a custom thing' and 'provide your special immutable handler' encompass - they could be as simple as a way of describing "you do Object x = yourTileObject.getSomething()" (fanciful ways of saying 'make an object variable that represents the return from .." and viola, you have the 'custom' immutable something for it, then and "then do yourTileObject.setBlah(x)" is the "provide your special immutable handler"....
OR it could be more involved (and likely is) in that it means "to store a string and an integer as custom data, you need to make a new class that has a string and integer in it, and the following 12 lines of boilerplate code required in all such new classes; then, you need to make a new class that extends that class and impliments XYZ , and has the following 12 boilerplate lines of code plus any additional getter/setters that you may need, then you use a game.GetRegistry().getsomebuilderthing(yournewclassname.class) and set the x and y values a,d viola, you now have your custom data system and can simply pass it in and get back the data with these following three lines "

Like the concept of making custom events being more than ' you make a class, put what you want in it, and that wiill work' it requires implmenting and extending certain base classes, including a bare-mininum of special methods and parameters in order to work with the system as a true event to pick up (Bukkit examples, no idea how sponge is similar or different, concept is likely similar, but way different pattern of code required)

Thus the request is for a walk-thru or sample working codes on what the process involves, what actual steps are required , what must those special things look like if new special classes must be made to be compatible with the infrastructure, with a working example to disect, practise with and then go "ah hah!"

Thank you.

**Comment by RobertHerhold** [comment]
So basically you want code examples to explain what is going on? Ideally it should be an explanation paired with code blocks... What page(s) are you looking at in particular?

**Comment by theboomer** [comment]
Wherever such pages may be best suited.  Not just an explanation (depending on what the steps are) but showing what minimum code needs to be added to meet the requirements  - again, bukkit event handler example - you can wave a hand and say "you just need to make your own event class and register it with x and you're good to go" or you can show how an example event class is made, which parts are required in all "empty" classes, and then "now add your objects here, add the getters here, and add this other code"  .. "Now, in your program, you evoke that event with x code, and any event handlers using that class of event for listeners will catch it"

so like an example "We're going to make a custom data unit that stores a string, a list of integers, and a uuid with a furnace block" showing what the classes need to be made in order to "make it so" and then of course, an example at the end of getting and setting it to round off.  

There are some of the comments that make it more 'a then b then c' pretty 'clear' but that could still be the equivalent of the form "just make your event class extend x and then register it and youre done".  There must be given things that may be more obvious so someone who can assemble the true big picture by looking at javadocs or the raw code but need to be made manefest in semi-spoonfeed mode.

**Comment by Inscrutable** [comment]
There are detailed code examples in a number of sections of the Docs, e.g. the Data API pages, with explanations around each. Is this the sort of thing you are asking for? Not every page has such lavish detail, because it takes time, talent, and dedication to get it done to a high standard.

If there are specific pages you feel need a more thorough treatment, or you want to propose new ones, could you offer some concrete examples?

Whilst I can understand your concern, and desire for more explanatory code examples, those who know it best are still hard at work getting Sponge ready for Beta. That has to take priority.
We are always looking for more contributors, so if you are up to the challenge, pick a topic and have a try at walking through it. Feel free to work up a Pull Request.

**Comment by saladoc** [comment]
@theboomer I invite you to take a look at the docs we currently have and tell us if the code examples we have fulfil your requirements of code examples. Currently you're only mentioning use cases that we have not covered on the docs yet and I'm unsure whether you are requesting those pages to be filled, criticising the extend of the code examples in the API so fare or wishing for more spoonfeeding on the forums.
Should it be the request, please refer to Inscrutables comment. 
Should it be the criticism, please provide more detail on a) which examples need to be more verbose and b) how verbose you would require them to be.
Should it be the wish regarding the forums, this is the wrong place to ask for it. While our docs staff are active on the forums and always eager to help or explain something, help there comes from more but just us, so it would better to take your request to the respective forum users that provide too abstract answers.

**Comment by theboomer** [comment]
Fine then...  instead of recognizing a _request_ it as a future-based, forward thinking planned nice-addition to keep in mind one day to add on, outlining a rather complicated process involving multiple stages and classes that need to be created to work together harmonously in a particular way, its considered a shot at the quality of the existing work in place.  

Clearly such a feature - one of the only singlemost largest biggest touted features of Sponge --  will _never_ be a good addition to put in when there is time for someone to request a fast working example written up by 'those who know' and 'those who can explain' are able to break it down "into high-standard quality work", cause it must be a perfect thing by "those who know" first shot out of the gate, and they dont have time to do that.

"Currently you only mentioning cases we have not yet covered" ... in my country, thats what "A request" is meant to address..

You can close this since its only dumb me unable to comprehend enough information to make a leap, i'll have to keep an eye out for plugins posted that might have cracked the process and steal their code when the time comes, and everyone else in the same boat just has to accept that "You just need to framistat the combobulon" conveys clearly all the specially required code and format necessary to fit in with the systems...

**Comment by theboomer** [comment]
Also, I thought I did provide a pretty concrete example in the form of "to store a string and an integer" in the first post, and in the second post, the actual sentence:
 like an example "We're going to make a custom data unit that stores a string, a list of integers, and a uuid with a furnace block" 

This is not the same as an existing example showing how to get the health data from an entity, or set the wet state of a sponge.  But since thats pretty unclear of an idea written as 'an example' then I have failed to communicate my idea, so dont worry about it...

**Comment by gabizou** [comment]
I think what they were trying to get at is, your first post is a bit wordy in stating what you're requesting. In this case, I can see now that you've been wanting clear documentation with code examples, of writing a custom `DataManipulator`. Am I correct?

**Comment by theboomer** [comment]
That is correct

**Comment by saladoc** [comment]
Code examples of varying verbosity are used throughout the API. Some are short, mere one-liners to demonstrate usage of a certain class (like the [Text Documentation](https://docs.spongepowered.org/en/plugin/text/text.html)), others deal with more implementation-heavy like the [Data Manipulator Contribution Docs](https://docs.spongepowered.org/en/contributing/implementation/datamanipulator.html).

Since I will most likely be the one to write the page you requested (see #335), I can tell you that there will be a dedicated page solely for the implementation of a custom `DataManipulator` which will be organized rather similar to the Data Contribution Docs: Basically the full code taken apart and explained. Would that format satisfy your request?

**Comment by theboomer** [comment]
It sounds like it would be on the order of what is required, given the multiple specific things to setup and minimum requirements they must meet in order to work together.

**Comment by saladoc** [comment]
Good. So if you do not object, I will treat this request as resolved for the time being since your request effectively mirrors what I had in mind for those docs.

If you take a look at our [pull requests](https://github.com/SpongePowered/SpongeDocs/pulls) regularly chances are high you will catch the pages while they are in progress. The Spongy bot provides a preview there for every pull request and you, as is every docs reader, are more than welcome to read it and give your feedback, be it about confusing wordings, typos in headings or an unsatisfactory amount of code examples. That way you should be able to respond quickly if the direction the pages are taking differ from what you think it should be.
